File: src-tauri/src/commands/claude.rs
use crate::application::claude_activity as claude_activity_app;
use crate::platform::{normalize_or_generate_trace_id, AppError, AppResult};

pub use crate::application::claude_activity::ClaudeActivityResponse;

fn classify_claude_error(message: &str) -> (&'static str, bool) {
    let lowered = message.to_lowercase();

    if lowered.contains("date must")
        || lowered.contains("invalid year")
        || lowered.contains("invalid month")
        || lowered.contains("invalid day")
        || lowered.contains("month must")
    {
        return ("invalid_input", false);
    }

    if lowered.contains("cannot find home directory") {
        return ("claude_home_not_found", false);
    }

    if lowered.contains("timed out") || lowered.contains("timeout") {
        return ("claude_timeout", true);
    }

    if lowered.contains("failed to read")
        || lowered.contains("permission denied")
        || lowered.contains("no such file")
    {
        return ("claude_filesystem_error", true);
    }

    ("claude_activity_failed", false)
}

fn map_claude_error(message: impl Into<String>, trace_id: &str, source: &str) -> AppError {
    let message = message.into();
    let (code, retryable) = classify_claude_error(&message);
    AppError::new(
        code,
        message,
        retryable,
        trace_id.to_string(),
        source.to_string(),
    )
}

#[tauri::command]
#[specta::specta]
pub async fn list_claude_projects(trace_id: Option<String>) -> AppResult<Vec<String>> {
    let trace_id = normalize_or_generate_trace_id(trace_id);
    claude_activity_app::list_claude_projects()
        .map_err(|error| map_claude_error(error, &trace_id, "claude.list_claude_projects.execute"))
}

#[tauri::command]
#[specta::specta]
pub async fn get_claude_activities(
    date: String,
    subscribed_folders: Vec<String>,
    trace_id: Option<String>,
) -> AppResult<ClaudeActivityResponse> {
    let trace_id = normalize_or_generate_trace_id(trace_id);
    claude_activity_app::get_claude_activities(date, subscribed_folders)
        .map_err(|error| map_claude_error(error, &trace_id, "claude.get_claude_activities.execute"))
}

#[tauri::command]
#[specta::specta]
pub async fn get_claude_activity_dates(
    subscribed_folders: Vec<String>,
    year: i32,
    month: u32,
    trace_id: Option<String>,
) -> AppResult<Vec<u32>> {
    let trace_id = normalize_or_generate_trace_id(trace_id);
    claude_activity_app::get_claude_activity_dates(subscribed_folders, year, month).map_err(
        |error| map_claude_error(error, &trace_id, "claude.get_claude_activity_dates.execute"),
    )
}

File: src/lib/claude.ts
import type { ClaudeActivityResponse } from "@/features/claude-activity/types";
import { invokeCommand } from "@/lib/platform";

export async function listClaudeProjects(): Promise<string[]> {
  return invokeCommand<string[]>("list_claude_projects", undefined, {
    fallbackMessage: "Failed to list Claude projects",
    source: "backend.claude.projects",
    code: "claude_projects_failed",
    traceArgName: "traceId",
  });
}

export async function getClaudeActivities(
  date: string,
  subscribedFolders: string[]
): Promise<ClaudeActivityResponse> {
  return invokeCommand<ClaudeActivityResponse>(
    "get_claude_activities",
    {
      date,
      subscribedFolders,
    },
    {
      fallbackMessage: "Failed to load Claude activity",
      source: "backend.claude.activities",
      code: "claude_activity_failed",
      traceArgName: "traceId",
    }
  );
}

export async function getClaudeActivityDates(
  subscribedFolders: string[],
  year: number,
  month: number
): Promise<number[]> {
  return invokeCommand<number[]>(
    "get_claude_activity_dates",
    {
      subscribedFolders,
      year,
      month,
    },
    {
      fallbackMessage: "Failed to load Claude activity dates",
      source: "backend.claude.activity_dates",
      code: "claude_activity_dates_failed",
      traceArgName: "traceId",
    }
  );
}
